### PR TITLE
GEM-40 | Implementation of "PUT /external/user-details" endpoint

### PR DESCRIPTION
**Pull request by pwgacek**

### CODE REVIEW

File: src/integration/kotlin/pl/edu/agh/gem/integration/controller/ExternalUserDetailsControllerIT.kt [hunk 1]
```diff
@@ -201,5 +211,69 @@ class ExternalUserDetailsControllerIT(
...
+            userDetailsRepository.save(createUserDetails(USER_ID))
+
+            // when
+            val response = service.updateGroupUserDetails(user, updateRequest)
+
+            // then
+            response shouldHaveHttpStatus OK
+        }
+
+        should("return NOT_FOUND when user details doesn't exist") {
```
Comment (lawos98) [inline src/integration/kotlin/pl/edu/agh/gem/integration/controller/ExternalUserDetailsControllerIT.kt]: that case can happen ? ( user dont have your own details )

File: src/main/kotlin/pl/edu/agh/gem/external/dto/validation/bankaccountnumber/NullOrBankAccountNumberPattern.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package pl.edu.agh.gem.external.dto.validation.bankaccountnumber
+
+import jakarta.validation.Constraint
+import kotlin.reflect.KClass
+
+@Target(AnnotationTarget.FIELD)
+@Retention(AnnotationRetention.RUNTIME)
+@MustBeDocumented
+@Constraint(validatedBy = [NullOrBankAccountNumberPatternValidator::class])
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/dto/validation/bankaccountnumber/NullOrBankAccountNumberPattern.kt]: do we want to add that to lib ?

**Review by lawos98 — COMMENTED** [review]
.

File: src/integration/kotlin/pl/edu/agh/gem/integration/controller/ExternalUserDetailsControllerIT.kt (hunk 1 above)
Comment (pwgacek) [reply src/integration/kotlin/pl/edu/agh/gem/integration/controller/ExternalUserDetailsControllerIT.kt]: zrobić zeby rzucało 5XX

File: src/integration/kotlin/pl/edu/agh/gem/integration/controller/ExternalUserDetailsControllerIT.kt [hunk 3]
```diff
@@ -112,15 +121,15 @@ class ExternalUserDetailsControllerIT(
             }
         }
 
-        should("return NOT_FOUND when user details doesn't exist") {
+        should("return INTERNAL_SERVER_ERROR when user details doesn't exist") {
```
Comment (lawos98) [inline src/integration/kotlin/pl/edu/agh/gem/integration/controller/ExternalUserDetailsControllerIT.kt]: could you edit test name to not have duplcates ?

File: src/main/kotlin/pl/edu/agh/gem/internal/service/UserDetailsService.kt [hunk 4]
```diff
@@ -22,6 +23,23 @@ class UserDetailsService(
...
+
+    fun updateUserDetails(userDetailsUpdate: UserDetailsUpdate) {
+        val userDetails = userDetailsRepository.findById(userDetailsUpdate.userId) ?: throw MissingUserDetailsException(userDetailsUpdate.userId)
+
+        val updatedUserDetails = UserDetails(
+            id = userDetails.id,
+            username = userDetailsUpdate.username ?: userDetails.username,
+            firstName = userDetailsUpdate.firstName ?: userDetails.firstName,
+            lastName = userDetailsUpdate.lastName ?: userDetails.lastName,
+            phoneNumber = userDetailsUpdate.phoneNumber ?: userDetails.phoneNumber,
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/internal/service/UserDetailsService.kt]: what update case when we want to remove phoneNumber or sth else?

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/internal/service/UserDetailsService.kt]: maybe when use doesnt't want to modify given field app will send null for this field. and if user wants to remove fielfd's value they pass empty string. What do you tihnk?

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/internal/service/UserDetailsService.kt]: second option is to always update - when new value is different ffrom old we put new value. when new value is the same as old one we also update, when new value is null we put null

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/internal/service/UserDetailsService.kt]: i prefer second option, what do u think?

Comment (lawos98) [reply src/main/kotlin/pl/edu/agh/gem/internal/service/UserDetailsService.kt]: sure we can do that in the second scenario

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/internal/service/UserDetailsService.kt]: done

File: src/test/kotlin/pl/edu/agh/gem/internal/service/UserDetailsServiceTest.kt [hunk 5]
```diff
@@ -95,4 +97,56 @@ class UserDetailsServiceTest : ShouldSpec({
...
+        val oldAttachmentId = "123"
+        val newFirstName = "John"
+        val newLastName = "Doe"
+
+        val userDetails = createBasicUserDetails(USER_ID, oldUserName, oldAttachmentId)
+        val userDetailsUpdate = createUserDetailsUpdate(
+            USER_ID,
+            newUserName,
+            newFirstName,
+            newLastName,
```
Comment (lawos98) [inline src/test/kotlin/pl/edu/agh/gem/internal/service/UserDetailsServiceTest.kt]: could you use in all places style like `id = USER_ID, username = newUserName` etc

Comment (pwgacek) [reply src/test/kotlin/pl/edu/agh/gem/internal/service/UserDetailsServiceTest.kt]: done

**Review by lawos98 — APPROVED** [review]

**Pull request merged**